### PR TITLE
ENH: adds --update-database option instead of update scripts

### PR DESCRIPTION
**Pull request by gethvi**
This is a PoC for consideration based on https://github.com/certtools/intelmq/issues/1517#issuecomment-621882632

I believe this is a better and more complete approach than having shell scripts, but it depends on what the community thinks.

Example cron:
`0  1  *  *  *  intelmq.bots.experts.tor_nodes.expert --update-database`

### CODE REVIEW

File: intelmq/bots/experts/tor_nodes/expert.py [hunk 1]
```diff
@@ -43,5 +50,48 @@ def process(self):
         self.send_message(event)
         self.acknowledge_message()
 
+    @classmethod
+    def run(cls):
+        if len(sys.argv) > 1 and sys.argv[1] == "--update-database":
```
Comment (wagner-certat) [inline intelmq/bots/experts/tor_nodes/expert.py:55]: As you maybe already imagined, I don't like that ;) The ideal solution would be a generic interface in the core which handles the argument parsing for this kind of things. However, I think that we can postpone this to future versions. It works reasonably well for now.

File: intelmq/bots/experts/tor_nodes/expert.py [hunk 2]
```diff
@@ -43,5 +50,48 @@ def process(self):
...
+            print(f"No bots of type {__name__} present in runtime.conf.")
+            exit(0)
+
+        try:
+            response = requests.get("https://check.torproject.org/exit-addresses")
+        except requests.exceptions.RequestException as e:
+            print(f"Connection Error: {e}")
+            exit(1)
+
+        ipv4_re = re.compile("(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)")
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/tor_nodes/expert.py]: I wanted to add some extra validation, perhaps needlessly. It's now updated to accept IPv6 too.

File: intelmq/bots/experts/tor_nodes/expert.py (hunk 1 above)
Comment (gethvi) [reply intelmq/bots/experts/tor_nodes/expert.py:55]: Yes I can imagine. :-) This was just the most straight forward solution for a working proof of concept code.

**Comment by gethvi** [comment]
Thanks for your suggestions, I fixed the code accordingly. I am happy you like this feature! I can work on the rest of the bots (asn-expert and geoip-expert) in context of this PR and then I can update the documentation as well. I am not sure what you had in mind in regards to migration (updating crontab with `intelmqctl upgrade-config`?).

**Comment by gethvi** [comment]
@wagner-certat All three bots are updated with --update-database option. I also tried to update the docs wherever I could find it. I believe the only remaining thing is the migration which I don't know how to do..

**Comment by gethvi** [comment]
Also this solves #1535 because it uses HTTP and not FTP.

**Comment by wagner-certat** [comment]
I prefer to postpone this change to the next feature release (2.3.0) as it has the potential of breaking things

**Comment by gethvi** [comment]
Ok. I will also add the check for #1057 in the meantime.

File: intelmq/bots/experts/tor_nodes/expert.py (hunk 1 above)
Comment (gethvi) [reply intelmq/bots/experts/tor_nodes/expert.py]: Should be improved now.

**Comment by gethvi** [comment]
This now truly solves #1535 because it uses HTTP (instead of FTP) and honors the `default.conf` settings (for proxy etc).

**Comment by gethvi** [comment]
There is a certain limitation with `argparse`, because it doesn't allow positional arguments (bot-id) not being present. Therefore it wouldn't be possible to call `intelmq.bots.. --update-database` without bot-id. I found a way around it, but it requires to call `sys.exit` after the `--update-database` has finished (so it doesn't try to use the string `"--update-database"` as a bot-id). Funnily, option like `--help` implements it's own `argparse.Action` and allows running without bot-id. But it does so in a similar manner using `sys.exit`. Anyway this is the reason, why `--update-database` option is placed inside mutually exclusive group of options. Because if there is ever going to be another option, calling both of them (e.g. `intelmq.bots.. --update-database --do-something-else`) would silently exit after whichever option was recognized first (based on the order of `if` inside `run`).

File: intelmq/lib/utils.py [hunk 3]
```diff
@@ -714,6 +718,36 @@ def send(self, *args, **kwargs):
         return super().send(*args, **kwargs)
 
 
+def create_request_session() -> requests.Session:
```
Comment (wagner-certat) [inline intelmq/lib/utils.py:721]: Looks similar to `intelmq.lib.bot.Bot.set_request_parameters` + `intelmq.lib.utils.create_request_session_from_bot`. Can we reduce the duplication a bit? The `set_request_parameters` could also be used statically I guess.

File: contrib/cron-jobs/update-asn-data [hunk 4]
```diff
@@ -1 +0,0 @@
-../../intelmq/bots/experts/asn_lookup/update-asn-data
```
Comment (wagner-certat) [inline contrib/cron-jobs/update-asn-data:1]: please leave it as is for backwards-compatibility

File: contrib/cron-jobs/update-geoip-data [hunk 5]
```diff
@@ -1 +0,0 @@
-../../intelmq/bots/experts/maxmind_geoip/update-geoip-data
```
Comment (wagner-certat) [inline contrib/cron-jobs/update-geoip-data:1]: please leave it as is for backwards-compatibility

File: contrib/cron-jobs/update-tor-nodes [hunk 6]
```diff
@@ -1 +0,0 @@
-../../intelmq/bots/experts/tor_nodes/update-tor-nodes
```
Comment (wagner-certat) [inline contrib/cron-jobs/update-tor-nodes:1]: please leave it as is for backwards-compatibility

File: intelmq/bots/experts/asn_lookup/expert.py [hunk 7]
```diff
@@ -57,5 +69,89 @@ def check(parameters):
...
+        bots = {}
+        runtime_conf = load_configuration(RUNTIME_CONF_FILE)
+        try:
+            for bot in runtime_conf:
+                if runtime_conf[bot]["module"] == __name__:
+                    bots[bot] = runtime_conf[bot]["parameters"]["database"]
+
+        except KeyError as e:
+            print("Database update failed. Your configuration of {0} is missing key {1}.".format(bot, e))
+            sys.exit(1)
```
Comment (wagner-certat) [inline intelmq/bots/experts/asn_lookup/expert.py:94]: combine with previous line, the the message is printed to stderr

File: intelmq/bots/experts/asn_lookup/expert.py [hunk 8]
```diff
@@ -57,5 +69,89 @@ def check(parameters):
...
+
+            url += str(months[0]) + "/RIBS/"
+            response = session.get(url)
+            pattern = re.compile(r"href=\"(rib\.\d{8}\.\d{4}\.bz2)\"")
+            days = pattern.findall(response.text)
+            days.sort(reverse=True)
+
+            if not days:
+                print("Database update failed. Couldn't find the latest database update.")
+                sys.exit(1)
```
Comment (wagner-certat) [inline intelmq/bots/experts/asn_lookup/expert.py:127]: same

File: debian/intelmq.install [hunk 9]
```diff
@@ -1,4 +1,4 @@
-debian/cron.d/intelmq-update-data etc/cron.d/
+debian/cron.d/intelmq-update-database etc/cron.d/
```
[the remark this answers is not in the document]
Comment (gethvi) [reply debian/intelmq.install:1]: To make it as consistent as possible.
* The commandline option is called `--update-database`
* The function behind it is `update_database`
* The datafile that is being updated is under the parameter `database` inside `runtime.conf`

When the user/administrator sees `intelmq-update-database` cronjob, it is easier to make the connection about what it does.

File: debian/cron.d/intelmq-update-database [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# /etc/cron.d/intelmq-update-database: crontab fragment for intelmq
+#  This updates the data files used by some expert bots.
+#
+# m h  dom mon dow      command
+# Update database for tor_nodes bot:
+11  0 * * *     intelmq   intelmq.bots.experts.tor_nodes.expert --update-database
+# Update database for maxmind_geoip bot:
+30 23 * * 2     intelmq   intelmq.bots.experts.maxmind_geoip.expert --update-database
+# Update database for asn_lookup bot:
+5 */2 * * *     intelmq   intelmq.bots.experts.asn_lookup.expert --update-database
```
[the remark this answers is not in the document]
Comment (gethvi) [reply debian/cron.d/intelmq-update-database:10]: That depends. But the source data are updated every two hours therefore having the default behavior of always using the most updated data seems reasonable to me.

File: intelmq/lib/bot.py [hunk 11]
```diff
@@ -60,6 +61,8 @@ class Bot(object):
     is_multithreadable = True
     # Collectors with an empty process() should set this to true, prevents endless loops (#1364)
     collector_empty_process = False
+    argparser = argparse.ArgumentParser(usage='%(prog)s [OPTIONS] BOT-ID')
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/lib/bot.py:64]: I did think about it and decided to place it outside the constructor for the following reasons:
* The content of ArgumentParser (subparsers, arguments, etc) is the same for all the instances of the bot which puts it in the position of class variable rather than instance variable. The ArgumentParser doesn't need to be instantiated with every bot instance.
* The main `run()` function is class method and the bot is only instantiated at the end of it therefore there wouldn't be available any ArgumentParser instance during the `run()` function when we need to parse the arguments.
* The `update_database` function was specifically built as a class method to be used in a fashion of a standalone script. It doesn't need to have an instance of the bot.

Now the real question is whether it is **desired to have any extra bot functionality in form of class methods** (== we do not need an instance of the bot) **or in the form of instance methods**. Both have it's pros and cons.

**Pros** (of having extra bot functionality as **instance methods**):
* Bot parameters and all of it's configuration is loaded and handled by the underlying class, you don't need to load and validate anything.

**Cons** (of having extra bot functionality as **instance methods**):
* If the bot is not fully configured it might fail when trying to instantiate the bot. For example if `update_database` was an instance method, it wouldn't be possible to call it without having correctly configured `pipeline.conf`, because `lib.bot.__init__` would fail on `lib.bot.__load_pipeline_configuration` which is rather silly considering that the pipeline has nothing to do with updating the database.
* You need to specify the bot id (to have the bot's configuration loaded), which is something that you might not want to do (or can't). This is supposed to be one of the benefits of this PR. Another example I can think of is (my) desire to use the `experts.sieve` to validate a filter syntax without the need to have the bot properly configured.

**Pros** (of having extra bot functionality as **class methods**):
* Removes cons of having extra functionality as instance methods (above).

**Cons** (of having extra bot functionality as **class methods**):
* Some code needs to be more or less duplicated (or perhaps have one code usable in both ways).
* Specific bot configuration might be ignored (in very exotic cases) because of missing bot-id. For example with `update_database` if we had two `experts.asn_lookup` bots with different proxy settings for some reason, the function has no clue which one to pick (currently it is completely ignored and the `defaults.conf` is used).

After considering all this I decided for this PR that having class method is the better approach. My long comment is also relevant to the next conversation regarding the code duplication in `def create_request_session() -> requests.Session:`. 

I welcome any insights on this.

File: intelmq/lib/utils.py (hunk 3 above)
Comment (gethvi) [reply intelmq/lib/utils.py:721]: The deeper reasoning behind this is explained above.

The code of `update_database` covers rare cases when there are more than one of the same experts in the pipeline. But with such consideration the code doesn't know which one's configuration (for proxy) to use. To make it predictable, it only uses `defaults.conf`. And for the class method approach (explained above) there is no method to load `requests.Session` object only based on `defaults.conf`. I don't see how to use `intelmq.lib.utils.create_request_session_from_bot` statically, it uses `self` for pretty much everything. :-(

Comment (gethvi) [reply intelmq/lib/utils.py]: I have updated the function `intelmq.lib.utils.create_request_session_from_bot` to a more universal approach (and renamed to reflect the change). Now it is optional to pass the bot instance to the function. If no bot is specified, only values from `defaults.conf`. Also I came across #1594.

File: intelmq/bots/experts/asn_lookup/expert.py [hunk 12]
```diff
@@ -57,5 +69,89 @@ def check(parameters):
...
+
+            print("Downloading the latest database update...")
+            url += days[0]
+            response = session.get(url)
+
+            if response.status_code != 200:
+                print("Database update failed. Server responded: {0}.".format(response.status_code))
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/asn_lookup/expert.py]: Done.

File: intelmq/bots/experts/asn_lookup/expert.py [hunk 13]
```diff
@@ -57,5 +69,89 @@ def check(parameters):
...
+                sys.exit(1)
+
+        except requests.exceptions.RequestException as e:
+            print("Database update failed. Connection Error: {0}".format(e))
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/asn_lookup/expert.py]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py [hunk 14]
```diff
@@ -72,5 +83,97 @@ def process(self):
...
+                    bots[bot] = runtime_conf[bot]["parameters"]["database"]
+
+        except KeyError as e:
+            print("Database update failed. Your configuration of {0} is missing key {1}.".format(bot, e))
+            if str(e) == "'license_key'":
+                print(
+                    "Since December 30, 2019 you need to register for a free license key to download GeoLite2 database.")
+                print(
+                    "https://blog.maxmind.com/2019/12/18/significant-changes-to-accessing-and-using-geolite2-databases/")
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/maxmind_geoip/expert.py]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py [hunk 15]
```diff
@@ -72,5 +83,97 @@ def process(self):
...
+        license_key = None
+        runtime_conf = load_configuration(RUNTIME_CONF_FILE)
+        try:
+            for bot in runtime_conf:
+                if runtime_conf[bot]["module"] == __name__:
+                    license_key = runtime_conf[bot]["parameters"]["license_key"]
+                    bots[bot] = runtime_conf[bot]["parameters"]["database"]
+
+        except KeyError as e:
+            print("Database update failed. Your configuration of {0} is missing key {1}.".format(bot, e))
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/maxmind_geoip/expert.py:109]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py [hunk 16]
```diff
@@ -72,5 +83,97 @@ def process(self):
...
+
+        if not bots:
+            print("Database update skipped. No bots of type {0} present in runtime.conf.".format(__name__))
+            sys.exit(0)
+
+        # we only need to import now, if there are no maxmind_geoip bots, this dependency does not need to be installed
+        try:
+            import maxminddb
+        except ImportError:
+            raise MissingDependencyError('geoip2')  # geoip2 depends on maxminddb
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/maxmind_geoip/expert.py]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py [hunk 17]
```diff
@@ -72,5 +83,97 @@ def process(self):
...
+            session = create_request_session()
+            response = session.get("https://download.maxmind.com/app/geoip_download",
+                                   params={
+                                       "license_key": license_key,
+                                       "edition_id": "GeoLite2-City",
+                                       "suffix": "tar.gz"
+                                   })
+        except requests.exceptions.RequestException as e:
+            print("Database update failed. Connection Error: {0}".format(e))
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/maxmind_geoip/expert.py]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py [hunk 18]
```diff
@@ -72,5 +83,97 @@ def process(self):
...
+
+        if response.status_code == 401:
+            print("Database update failed. Your license key is invalid.")
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/maxmind_geoip/expert.py]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py [hunk 19]
```diff
@@ -72,5 +83,97 @@ def process(self):
...
+
+        if response.status_code != 200:
+            print("Database update failed. Server responded: {0}.".format(response.status_code))
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/maxmind_geoip/expert.py]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py [hunk 20]
```diff
@@ -72,5 +83,97 @@ def process(self):
...
+            with tarfile.open(fileobj=io.BytesIO(response.content), mode='r:gz') as archive:
+                for member in archive.getmembers():
+                    if "GeoLite2-City.mmdb" in member.name:
+                        database_data = maxminddb.open_database(database=archive.extractfile(member),
+                                                                mode=maxminddb.MODE_FD)
+                        break
+
+        except maxminddb.InvalidDatabaseError:
+            print("Database update failed. Database file invalid.")
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/maxmind_geoip/expert.py]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py [hunk 21]
```diff
@@ -72,5 +83,97 @@ def process(self):
...
+
+        if not database_data:
+            print("Database update failed. Could not locate file 'GeoLite2-City.mmbd' in the downloaded archive.")
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/maxmind_geoip/expert.py]: Done.

File: intelmq/bots/experts/tor_nodes/expert.py [hunk 22]
```diff
@@ -43,5 +55,60 @@ def process(self):
...
+        bots = {}
+        runtime_conf = load_configuration(RUNTIME_CONF_FILE)
+        try:
+            for bot in runtime_conf:
+                if runtime_conf[bot]["module"] == __name__:
+                    bots[bot] = runtime_conf[bot]["parameters"]["database"]
+
+        except KeyError as e:
+            print("Database update failed. Your configuration of {0} is missing key {1}.".format(bot, e))
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/tor_nodes/expert.py]: Done.

File: intelmq/bots/experts/tor_nodes/expert.py [hunk 23]
```diff
@@ -43,5 +55,60 @@ def process(self):
...
+            print("Database update skipped. No bots of type {0} present in runtime.conf.".format(__name__))
+            sys.exit(0)
+
+        try:
+            print("Downloading the latest database update...")
+            session = create_request_session()
+            response = session.get("https://check.torproject.org/exit-addresses")
+        except requests.exceptions.RequestException as e:
+            print("Database update failed. Connection Error: {0}".format(e))
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/tor_nodes/expert.py]: Done.

File: intelmq/bots/experts/tor_nodes/expert.py [hunk 24]
```diff
@@ -43,5 +55,60 @@ def process(self):
...
+
+        if response.status_code != 200:
+            print("Database update failed. Server responded: {0}.".format(response.status_code))
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/tor_nodes/expert.py]: Done.

File: intelmq/bots/experts/asn_lookup/expert.py (hunk 7 above)
Comment (gethvi) [reply intelmq/bots/experts/asn_lookup/expert.py]: Done.

File: intelmq/bots/experts/asn_lookup/expert.py (hunk 8 above)
Comment (gethvi) [reply intelmq/bots/experts/asn_lookup/expert.py]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py [hunk 25]
```diff
@@ -72,5 +83,97 @@ def process(self):
...
+            if str(e) == "'license_key'":
+                print(
+                    "Since December 30, 2019 you need to register for a free license key to download GeoLite2 database.")
+                print(
+                    "https://blog.maxmind.com/2019/12/18/significant-changes-to-accessing-and-using-geolite2-databases/")
+            sys.exit(1)
+
+        if not bots:
+            print("Database update skipped. No bots of type {0} present in runtime.conf.".format(__name__))
+            sys.exit(0)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/maxmind_geoip/expert.py:118]: This case I wouldn't consider an error. The script finishes successfully, it just doesn't have anything to do.

File: intelmq/bots/experts/asn_lookup/expert.py [hunk 26]
```diff
@@ -57,5 +69,89 @@ def check(parameters):
...
+                print("Database update failed. Couldn't find the latest database update.")
+                sys.exit(1)
+
+            print("Downloading the latest database update...")
+            url += days[0]
+            response = session.get(url)
+
+            if response.status_code != 200:
+                print("Database update failed. Server responded: {0}.".format(response.status_code))
+                sys.exit(1)
```
[the remark this answers is not in the document]
Comment (gethvi) [reply intelmq/bots/experts/asn_lookup/expert.py]: Done.

File: contrib/cron-jobs/update-asn-data (hunk 4 above)
Comment (gethvi) [reply contrib/cron-jobs/update-asn-data]: Done.

File: contrib/cron-jobs/update-geoip-data (hunk 5 above)
Comment (gethvi) [reply contrib/cron-jobs/update-geoip-data]: Done.

File: contrib/cron-jobs/update-tor-nodes (hunk 6 above)
Comment (gethvi) [reply contrib/cron-jobs/update-tor-nodes]: Done.

File: intelmq/bots/experts/maxmind_geoip/expert.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (wagner-certat) [reply intelmq/bots/experts/maxmind_geoip/expert.py:118]: ah, sorry, of course

**Comment by gethvi** [comment]
I have added also `update_database` for `experts.recordedfuture_iprisk`, however I do not have a valid API token to properly test it.

File: intelmq/lib/bot.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (wagner-certat) [reply intelmq/lib/bot.py:64]: Thanks for your thorough response, I did by far not expect such a detailed explanation on such a short comment.

tl;dr I'm totally ok with leaving it there.

>     * The content of ArgumentParser (subparsers, arguments, etc) is the same for all the instances of the bot which puts it in the position of class variable rather than instance variable. The ArgumentParser doesn't need to be instantiated with every bot instance.

I'm not sure if that matters somewhere (I.e. we create a bot instance and the constructor is not called).

> Now the real question is whether it is **desired to have any extra bot functionality in form of class methods** (== we do not need an instance of the bot) **or in the form of instance methods**. Both have it's pros and cons.

We may end up with both. We now have class methods because it is easier or at all possible (for instance methods we block the actually running bot instance)

>     * Some code needs to be more or less duplicated (or perhaps have one code usable in both ways).

I prefer the latter

>     * Specific bot configuration might be ignored (in very exotic cases) because of missing bot-id. For example with `update_database` if we had two `experts.asn_lookup` bots with different proxy settings for some reason, the function has no clue which one to pick (currently it is completely ignored and the `defaults.conf` is used).

This is an issue, but can be solved by (more) code.

I agree with everything else.

**Comment by wagner-certat** [comment]
Unfortunately, the tests are failing now: https://travis-ci.org/github/certtools/intelmq/jobs/716521834#L1031 (also locally for me). Otherwise I'd merge it

**Comment by wagner-certat** [comment]
Thank you so much @gethvi :)

Some further adaptions from my side:
852a1c2b05e1b3446c8fe8d54aa8b4ebb8b4ac89 8b6bac6beb7d6dcb2089f79d9bfc74520475fa7b 0582566dfa9f67b69dc8873e1c77d7597aa32f34 34db562a01b4a42c283e961bb7d29333f46ff315
